File: Labs/Lab 6/Python/CanFinishCourses.py
from collections import deque
class CanFinishCourses:
    def canFinish(self, numCourses, prerequisites):
        indegree = [0]*numCourses
        for prereq in prerequisites:
            indegree[prereq[0]] += 1
        queue = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        result = []
        while queue:
            node = queue.popleft()
            result.append(node)
            for prereq in prerequisites:
                if prereq[1] == node:
                    indegree[prereq[0]] -= 1
                    if indegree[prereq[0]] == 0:
                        queue.append(prereq[0])
        if len(result) != numCourses:
            return False
        return True

File: Labs/Lab 6/Java/CanFinishCourses.java
package Lab6;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class CanFinishCourses { 
	public boolean canFinish(int numCourses, int[][] prerequisites) {
		int[] indegree = new int[numCourses];
		for(int[] prereq: prerequisites) {
				indegree[prereq[0]]++;
		}
		Queue<Integer> q = new LinkedList<>();
		for(int i = 0; i < numCourses; i++) {
			if(indegree[i] == 0) {
				q.add(i);
			}
		}
		List<Integer> result = new ArrayList<>();
		while(!q.isEmpty()) {
			int node = q.poll();
			result.add(node);
			for (int[] prereq : prerequisites) {
	            if (prereq[1] == node) {
	                indegree[prereq[0]]--;
	                if (indegree[prereq[0]] == 0) {
	                    q.add(prereq[0]);
	                }
	            }
			}
		}
		if (result.size()!= numCourses) {
			return false;
		}
		return true;
	}
	public static void main(String[] args) {
		CanFinishCourses c = new CanFinishCourses();
		int numCourses = 2;
		int[][] prerequisites = {{1,0}};
		System.out.println(c.canFinish(numCourses, prerequisites));
		CanFinishCourses c2 = new CanFinishCourses();
		int numCourses2 = 2;
		int[][] prerequisites2 = {{1,0},{0,1}};
		System.out.println(c2.canFinish(numCourses2, prerequisites2));
	}
}
